File: src/infrastructure/tools/file/write-file.ts
/**
 * write_file — 写入（整文件覆盖）文件
 *
 * 自动创建父目录；返回归一化后内容的 hash10，可直接作为后续 edit_file 的 expected_hash。
 */
import * as tauriFs from '@tauri-apps/plugin-fs'
import { t } from '@/ui/i18n'
import { toolRegistry } from '@/domain/tools'
import type { ToolContext, ToolExecutor, ToolResult } from '@/domain/tools/types'
import { securityService } from '@/services/security-service'
import { computeContentHash10, formatSize, isTauriFsAvailable } from './common'

toolRegistry.register(
    'write_file',
    (async (args: Record<string, any>, ctx: ToolContext): Promise<ToolResult> => {
    if (!isTauriFsAvailable())
      throw '[write_file] Error: not running in a Tauri environment'

    const fullPath = await securityService.resolveSafePath(
      args.path as string,
      'w',
      ctx.sessionId,
    )
    const content = args.content as string
    try {
      // 创建父目录（兼容 Windows 反斜杠路径）
      const normalizedPath = fullPath.replace(/\\/g, '/')
      const parent = normalizedPath.substring(0, normalizedPath.lastIndexOf('/'))
      if (parent) {
        await tauriFs.mkdir(parent, { recursive: true }).catch(() => {})
      }
      const existed = await tauriFs.exists(fullPath).catch(() => false)
      await tauriFs.writeTextFile(fullPath, content)

      // 计算归一化内容的 hash10，与 read_file/edit_file 一致
      const hash10 = await computeContentHash10(content)
      const lineCount = content.replace(/\r\n/g, '\n').split('\n').length
      const size = formatSize(new TextEncoder().encode(content).length)

      // ⚠️ 模型侧固定英文（P4b/D2-A）。
      // 文案与 Rust `native_tools/file/write_file.rs` 一致（铁律 1）。
      const returnContent = existed
        ? `✅ File overwritten (${size}): ${fullPath}`
        : `✅ File created (${size}): ${fullPath}`

      return {
        uiData: {
          hash10,
          fullPath,
          lineCount,
          byteSize: new TextEncoder().encode(content).length,
        },
        content: returnContent + `\n🔑 hash10: ${hash10}`,
      }
    } catch (e: any) {
      throw `Error: failed to write file — ${e.message || String(e)}`
    }
  }) as ToolExecutor,
    t('写入文件'),
)

File: src-tauri/virlen-core/src/agent/native_tools/file/write_file.rs
//! `write_file` 工具（原生）— 全量写入（覆写或创建），返回 hash10 与行/字节统计。

use crate::agent::native_tools::common::{arg_str, resolve_safe_path};
use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use crate::file_ops;
use serde_json::{json, Value};

use super::common::format_size;

pub(crate) async fn write_file_tool(
    ctx: &NativeToolCtx<'_>,
    args: &Value,
) -> Result<NativeToolOutcome, String> {
    let path = arg_str(args, "path").unwrap_or_default();
    let full_path = resolve_safe_path(&path, "w", ctx.security)?;
    let content = arg_str(args, "content").unwrap_or_default();

    let result = {
        let full_path_c = full_path.clone();
        let content_c = content.clone();
        tokio::task::spawn_blocking(move || file_ops::write_file(&full_path_c, &content_c))
            .await
            .map_err(|e| format!("Task join error: {}", e))?
            .map_err(|e| format!("Error: failed to write file — {}", e))?
    };

    let existed = result.existed;
    let return_content = if existed {
        format!("✅ File overwritten ({}): {}", format_size(result.byte_size), full_path)
    } else {
        format!("✅ File created ({}): {}", format_size(result.byte_size), full_path)
    };

    Ok(NativeToolOutcome::Value {
        content: format!("{}\n🔑 hash10: {}", return_content, result.hash10),
        ui_data: Some(json!({
            "hash10": result.hash10,
            "fullPath": full_path,
            "lineCount": result.line_count,
            "byteSize": result.byte_size,
        })),
    })
}
